Add docs on custom Operations (#3507)
airbytehq/airbyte: docs/tutorials/transformation-and-normalization/transformations-with-dbt.md

# Transformations with DBT \(Part 2/3\)

## Overview

This tutorial will describe how to integrate SQL based transformations with Airbyte syncs using specialized transformation tool: DBT.

> This tutorial is the second part of the previous tutorial [Connecting EL with T using SQL](transformations-with-sql.md). Next, we'll wrap-up with a third part on submitting transformations back in Airbyte: [connecting EL with T in Airbyte](transformations-with-airbyte.md).

Reviewer: ```suggestion
This tutorial is the second part of the previous tutorial [Transformations with dbt](transformations-with-sql.md). Next, we'll wrap-up with a third part on submitting transformations back in Airbyte: [Transformations with Airbyte](transformations-with-airbyte.md).
```

(Example outputs are updated with Airbyte version 0.23.0-alpha from May 2021)

## Transformations with dbt

The tool in charge of transformation behind the scenes is actually called [DBT](https://blog.getdbt.com/what--exactly--is-dbt-/) \(Data Build Tool\).

> Before generating the SQL files as we've seen previously, Airbyte is setting up internally a Docker image where DBT is installed, and automatically generates a DBT project which is created as specified in the [dbt projet documentation page](https://docs.getdbt.com/docs/building-a-dbt-project/projects) with the right credentials for the target destination. The DBT models are then run afterward, thanks to [DBT CLI](https://docs.getdbt.com/dbt-cli/cli-overview).

Reviewer: ```suggestion
Before generating the SQL files as we've seen in the previous tutorial, Airbyte sets up a dbt Docker instance and automatically generates a dbt project for us. This is created as specified in the [dbt project documentation page](https://docs.getdbt.com/docs/building-a-dbt-project/projects) with the right credentials for the target destination. The dbt models are then run afterward, thanks to the [dbt CLI](https://docs.getdbt.com/dbt-cli/cli-overview).
```

> However, for now, let's see how to interact with the DBT tool.

Reviewer: ```suggestion
However, for now, let's run through working with the dbt tool.
```

### Validate dbt project settings

> Let's say we identified our workspace (as shown in the previous tutorial [Connecting EL with T using SQL](transformations-with-sql.md)), and we have a workspace ID of:

Reviewer: ```suggestion
Let's say we identified our workspace (as shown in the previous tutorial [Transformations with SQL](transformations-with-sql.md)), and we have a workspace ID of:
```

This tutorial is the second part of the previous tutorial [Connecting EL with T using SQL](transformations-with-sql.md).

(Example outputs are updated with Airbyte version 0.23.0-alpha from May 2021)

## Transformations with dbt

The tool in charge of transformation behind the scenes is actually called [DBT](https://blog.getdbt.com/what--exactly--is-dbt-/) \(Data Build Tool\).

Before generating the SQL files as we've seen previously, Airbyte is setting up internally a Docker image where DBT is installed, and automatically generates a DBT project which is created as specified in the [dbt projet documentation page](https://docs.getdbt.com/docs/building-a-dbt-project/projects) with the right credentials for the target destination. The DBT models are then run afterward, thanks to [DBT CLI](https://docs.getdbt.com/dbt-cli/cli-overview).

> m

Reviewer 2: ```suggestion

```

  keepalives_idle: 0
  sslmode: None
  Connection test: OK connection ok

> ```

Author: Yes, everything should be 'OK'

It's mainly to see the expected output of the dbt command look like too

-- covid_epidemiology

{{ config(alias="covid_epidemiology_ab3", schema="_airbyte_quarantine", tags=["top-level-intermediate"]) }}
-- SQL model to build a hash column based on the values of this record
select
    *,
    {{ dbt_utils.surrogate_key([
        adapter.quote('key'),
        adapter.quote('date'),
        'new_tested',
        'new_deceased',
        'total_tested',
        'new_confirmed',
        'new_recovered',
        'total_deceased',
        'total_confirmed',
        'total_recovered',
    ]) }} as _airbyte_covid_epidemiology_hashid
from {{ ref('covid_epidemiology_ab2_558') }}
-- covid_epidemiology

{{ config(alias="covid_epidemiology", schema="quarantine", tags=["top-level"]) }}
-- Final base SQL model
select
    {{ adapter.quote('key') }},
    {{ adapter.quote('date') }},
    new_tested,
    new_deceased,
    total_tested,
    new_confirmed,
    new_recovered,
    total_deceased,
    total_confirmed,
    total_recovered,
    _airbyte_emitted_at,
    _airbyte_covid_epidemiology_hashid
from {{ ref('covid_epidemiology_ab3_558') }}
-- covid_epidemiology from {{ source('quarantine', '_airbyte_raw_covid_epidemiology') }}
```

> If you have [DBT installed](https://docs.getdbt.com/dbt-cli/installation/) locally on your machine, you can then view, edit, version, customize, and run the dbt models in your project outside Airbyte syncs.

Reviewer: ```suggestion
If you have [dbt installed](https://docs.getdbt.com/dbt-cli/installation/) locally on your machine, you can then view, edit, version, customize, and run the dbt models in your project outside Airbyte syncs.
```

Example Output:

```text
Running with dbt=0.19.1
Installing https://github.com/fishtown-analytics/dbt-utils.git@0.6.4
  Installed from revision 0.6.4

Running with dbt=0.19.1
Found 4 models, 0 tests, 0 snapshots, 0 analyses, 364 macros, 0 operations, 0 seed files, 1 source, 0 exposures

Concurrency: 32 threads (target='prod')

1 of 1 START table model quarantine.covid_epidemiology....................................................... [RUN]
1 of 1 OK created table model quarantine.covid_epidemiology.................................................. [SELECT 35822 in 0.44s]

Finished running 1 table model in 0.63s.

Completed successfully

Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

> Now, that you've exported the generated normalization models, you can decide to edit and tweak them differently...

Reviewer: ```suggestion
Now, that you've exported the generated normalization models, you can edit and tweak them as necessary.
```

Example Output:

```text
Running with dbt=0.19.1
Installing https://github.com/fishtown-analytics/dbt-utils.git@0.6.4
  Installed from revision 0.6.4

Running with dbt=0.19.1
Found 4 models, 0 tests, 0 snapshots, 0 analyses, 364 macros, 0 operations, 0 seed files, 1 source, 0 exposures

Concurrency: 32 threads (target='prod')

1 of 1 START table model quarantine.covid_epidemiology....................................................... [RUN]
1 of 1 OK created table model quarantine.covid_epidemiology.................................................. [SELECT 35822 in 0.44s]

Finished running 1 table model in 0.63s.

Completed successfully

Done. PASS=1 WARN=0 ERROR=0 SKIP=0 TOTAL=1
```

> ### Hooking up a custom dbt project to Airbyte syncs

Reviewer 2: this feels out of place. I'm also slightly lost when reading this without some UI reference. this feels like it should either be a separate tutorial with pictures or part of the operations.md doc describing how to configure a private git repo.

Author: Ok i can make a part 3 tutorial then! (once I have a UI for pictures)
